### PR TITLE
move set_nonblocking() to server with OptionalSslAcceptor

### PR DESCRIPTION
**Pull request by poidl**
Method `set_nonblocking()` is only implemented for servers with `SslAcceptor`, although its implementation is also valid for non-SSL servers. This moves the method to the non-specific server implementations, so it can be used for both.

### CODE REVIEW

**Comment by illegalprime** [comment]
Thanks for the PR! Could you add tests? And methods to get access to the underlying TcpListener? The tests are just to make sure it works on both setups.

**Comment by poidl** [comment]
Oh, I totally forgot about the tests, also to execute cargo test. Okay I'll
work on it next weekend!

(But in case somebody needs the fix integrated earlier, then please go
ahead, no need to wait for me)

**Comment by illegalprime** [comment]
@poidl be sure to run `cargo fmt` in the toplevel of the project so all the formatting stays the same, that's why the build is failing right now. (If you don't have rust format, you can install it with `cargo install rustfmt`)

**Comment by poidl** [comment]
This one passes. Not sure if the doc example makes sense, there may be more typical uses for a non-blocking server. Feel free to change. Also, when I executed `cargo fmt` in the root directory, it changed a couple of files I hadn't touched, which caused the previous CI test to fail. Running it only on the modified file worked.

File: src/server/mod.rs [hunk 1]
```diff
@@ -271,3 +305,43 @@ impl Iterator for Server<NoSslAcceptor> {
...
+					Ok(t) => t,
+					Err(e) => panic!("received error for `{}`: {}", stringify!($e), e),
+				}
+			}
+		}
+		use super::*;
+
+		// Test unsecure server
+
+		let mut server = t!(Server::bind("127.0.0.1:0"));
```
Comment (illegalprime) [inline src/server/mod.rs]: You can just call `.unwrap()` instead of use the macro.

File: src/server/mod.rs [hunk 2]
```diff
@@ -271,3 +305,43 @@ impl Iterator for Server<NoSslAcceptor> {
...
+		// will block indefinitely.
+		t!(server.set_nonblocking(true));
+
+		let result = server.accept();
+		match result {
+			// nobody tried to establish a connection, so we expect an error
+			Ok(_) => panic!("expected error"),
+			Err(e) => {
+				match e.error {
+					HyperIntoWsError::Io(e) => if e.kind() == io::ErrorKind::WouldBlock {},
```
Comment (illegalprime) [inline src/server/mod.rs]: What this match arm is doing is matching against all `HyperIntoWsError::Io(e)` errors, then it checks if it is a `WouldBlock` error and does nothing either way. If you want to use a match guard you would do:
```rust
HyperIntoWsError::Io(e) if e.kind() == io::ErrorKind::WouldBlock =>
```

and not 

```rust
HyperIntoWsError::Io(e) => if e.kind() == io::ErrorKind::WouldBlock
```

It's kind of subtle.

File: src/server/mod.rs [hunk 3]
```diff
@@ -271,3 +305,43 @@ impl Iterator for Server<NoSslAcceptor> {
...
+
+		let mut server = t!(Server::bind("127.0.0.1:0"));
+
+		// Note that if set_nonblocking() doesn't work, but the following
+		// fails to panic for some reason, then the .accept() method below
+		// will block indefinitely.
+		t!(server.set_nonblocking(true));
+
+		let result = server.accept();
+		match result {
```
Comment (illegalprime) [inline src/server/mod.rs]: Might as well match on `result.unwrap_err()` if you're going to panic with an error.

File: src/server/mod.rs [hunk 4]
```diff
@@ -271,3 +305,43 @@ impl Iterator for Server<NoSslAcceptor> {
...
+		use super::*;
+
+		// Test unsecure server
+
+		let mut server = t!(Server::bind("127.0.0.1:0"));
+
+		// Note that if set_nonblocking() doesn't work, but the following
+		// fails to panic for some reason, then the .accept() method below
+		// will block indefinitely.
+		t!(server.set_nonblocking(true));
```
Comment (illegalprime) [inline src/server/mod.rs]: Again just use `.unwrap()`, remove the `t` macro.

File: src/server/mod.rs [hunk 5]
```diff
@@ -156,6 +156,48 @@ impl<S> Server<S>
...
+	/// // our websocket server, and if so, handle the TcpStream.
+	/// // (2) In between we need to something else, possibly unrelated to networking.
+	///
+	/// let mut server = Server::bind("127.0.0.1:0").unwrap();
+	///
+	/// // Set the server to non-blocking.
+	/// server.set_nonblocking(true);
+	///
+	/// for i in 1..3 {
+	/// 	let result = server.accept();
```
Comment (illegalprime) [inline src/server/mod.rs]: nit: I would merge this line and the next.

File: src/server/mod.rs (hunk 2 above)
Comment (poidl) [inline src/server/mod.rs]: Oops, that's a logical error. Thanks.

File: src/server/mod.rs (hunk 3 above)
Comment (poidl) [inline src/server/mod.rs]: This doesn't compile:
```
error: no method named `unwrap_err` found for type `std::result::Result<server::upgrade::WsUpgrade<std::net::TcpStream>, server::InvalidConnection<std::net::TcpStream>>` in the current scope
   --> src/server/mod.rs:327:16
    |
327 |           match result.unwrap_err() {
    |                        ^^^^^^^^^^
    |
    = note: the method `unwrap_err` exists but the following trait bounds were not satisfied: `server::upgrade::WsUpgrade<std::net::TcpStream> : std::fmt::Debug`
```

File: src/server/mod.rs (hunk 4 above)
Comment (poidl) [inline src/server/mod.rs]: done

File: src/server/mod.rs (hunk 1 above)
Comment (poidl) [inline src/server/mod.rs]: done

File: src/server/mod.rs (hunk 5 above)
Comment (poidl) [inline src/server/mod.rs]: done

File: src/server/mod.rs (hunk 3 above)
Comment (illegalprime) [inline src/server/mod.rs]: Oh yeah, then that's fine

**Comment by poidl** [comment]
Awesome, thanks!

**Comment by illegalprime** [comment]
Thank you for all the work you put into this @poidl !

**Comment by poidl** [comment]
Thanks for the guidance, I feel like a learned a lot!

**Pull request merged**